core/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/dao/CommonPermissionStorage.java

```diff
@@ -0,0 +1,205 @@
+
+import org.bson.Document;
+import org.eclipse.che.api.core.ConflictException;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.mongodb.client.model.Filters.and;
+import static com.mongodb.client.model.Filters.eq;
+import static com.mongodb.client.model.Filters.in;
+import static java.util.function.Function.identity;
+
+/**
+ * Common implementation for {@link PermissionsStorage} based on MongoDB storage.
+ *
+ * <p>Stores permissions of domains that bound by {@link CommonDomains}
+ *
+ * <p>Example of domain binding
+ * <pre>
+ *     Multibinder<PermissionsDomain> multibinder = Multibinder.newSetBinder(binder(), PermissionsDomain.class, CommonDomains.class);
+ *     multibinder.addBinding().toInstance(new PermissionsDomain("myDomain",
+ *                                                               new HashSet&lt;&gt;(Arrays.asList("read", "write", "use", "setPermissions"))));
+ * </pre>
+ *
+ * <p>Permissions collection document scheme:
+ * <pre>
+ *
+ * {
+ *     "user" : "user123",
+ *     "domain" : "workspace",
+ *     "instance" : "workspace123",
+ *     "actions" : [
+ *         "read",
+ *         "write",
+ *         ...
+ *     ]
+ * }
+ *
+ * </pre>
+ *
+ * @author Sergii Leschenko
+ */
+@Singleton
+public class CommonPermissionStorage implements PermissionsStorage {
+    private final MongoCollection<PermissionsImpl> collection;
+
+    private final Map<String, PermissionsDomain> idToDomain;
+
+    @Inject
+    public CommonPermissionStorage(@Named("mongo.db.organization") MongoDatabase database,
+                                   @Named("organization.storage.db.permission.collection") String collectionName,
+                                   @CommonDomains Set<PermissionsDomain> permissionsDomains) throws IOException {
+        collection = database.getCollection(collectionName, PermissionsImpl.class);
+        collection.createIndex(new Document("user", 1).append("domain", 1).append("instance", 1), new IndexOptions().unique(true));
+
+        this.idToDomain = permissionsDomains.stream()
+                                            .collect(Collectors.toMap(PermissionsDomain::getId, identity()));
+    }
+
+    @Override
+    public Set<PermissionsDomain> getDomains() {
+        return new HashSet<>(idToDomain.values());
+    }
+
+    @Override
+    public void store(PermissionsImpl permissions) throws ServerException {
+        final PermissionsDomain permissionsDomain = idToDomain.get(permissions.getDomain());
+        if (permissionsDomain == null) {
+            throw new IllegalArgumentException("Storage doesn't support domain with id '" + permissions.getDomain() + "'");
```

Comment:
My motivation is next: Users' calls should be processed by PermissionManager. And it should throw ConflictException in case of requesting unsupported domain. IllegalArgumentException can be thrown only if somebody use CommonPermissionStorage directly and try to store unsupported domain. WDYT?